text/0240-app-telemetry-schema-vision-roadmap.md

```diff
@@ -0,0 +1,429 @@
+source code of the applications and their components. The current stack of
+OpenTelemetry follows this approach and offers a unified mechanism that allows
+this ecosystem to report telemetry data jointly via a generic client interface
+and a common protocol through an SDK. Moreover, OpenTelemetry's semantic
+conventions establish a vendor-agnostic and standardized vocabulary to ensure
+more consistent instrumentation over time. This standardization facilitates the
+analysis and utilization of these metadata data across the entire telemetry
+pipeline.
+
+But this approach is not without challenges:
+
+* **Discoverability and Interoperability**: It is difficult to discover a priori
+  and automatically what an application as a whole specifically generates in terms
+  of telemetry and associated metadata.This makes it difficult to integrate with
+  enterprise data catalogs, compliance procedures, or automated privacy
+  enforcement in _CI/CD pipelines_.
+* **User experience**: Although very flexible, generic clients do not offer the
+  same benefits as a strongly typed dedicated API. A type-safe API is more
+  ergonomic, more robust, and more easily maintainable. Modern IDEs are capable
+  of providing smart autocompletion and contextual documentation based on the API.
+  Compilers can automatically detect errors in case of refactoring or evolution
+  of the telemetry schema.
+* **Extensibility**: Adding metadata to the basic signal specification is
+  essential for enabling use cases like data security, privacy enforcement,
+  metadata-driven data transformation, and knowledge graph enrichment. Currently,
+  there's no standard way to add metadata separate from an application's or
+  library's development cycle. These metadata definitions should be distinct from
+  the signal definitions and could be specified by teams other than development
+  teams.
+* **Performance overheads**: A significant downside of generic telemetry
+  instrumentation is the various overheads it generally introduces due to inherent
+  layers of abstraction. For example, the collection of attributes is typically
+  represented as a list of key/value pairs or as hashmaps, resulting in memory
+  overhead. A simple struct or a set of well-typed function arguments will be more
+  efficient and less error-prone for representing this list of attributes. In the
+  same way, it is possible to use a dictionary encoding for values whose domain is
+  specified in the form of an enumeration in the schema definition.
+
+Databases and RPC systems (e.g., Protobuf & gRPC) have already addressed some of
+these issues with a schema-first approach. There is nothing to prevent adopting
+a similar approach in the context of telemetry. **This document discusses how to
+apply a schema-first approach in the OpenTelemetry project and its implications
+for the existing Telemetry Schema and Semantic Conventions.**
+
+The following diagram provides a conceptual overview of the relationships
```

Comment:
this still goes under Current State header, a bit confusing, maybe add "Possible/Desired Future State"